File: src/lib/evm/sponsor/global_paymaster.ts
/**
 * Encode `paymasterAndData` for `PactoGlobalPaymaster` (ERC-4337 EntryPoint v0.7).
 *
 * Layout:
 *   [0:20]  paymaster
 *   [20:36] uint128 verificationGasLimit
 *   [36:52] uint128 postOpGasLimit
 *   [52:]   abi.encode(uint8 version, bytes32 npubHash, address member, address policy)
 *
 * Member path: `policy` must be the zero address. Vectors: `./globalPaymasterAndData.vectors.json`.
 */

export const GLOBAL_PAYMASTER_DATA_VERSION = 1 as const;
export const GLOBAL_PAYMASTER_DATA_OFFSET = 52 as const;
export const GLOBAL_BALANCE_HEADROOM_BPS = 11_500 as const;
export const DEFAULT_GLOBAL_PAYMASTER_VERIFICATION_GAS_LIMIT = 500_000 as const;
export const DEFAULT_GLOBAL_POST_OP_GAS_LIMIT = 50_000 as const;
export const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000' as const;

export type Address = `0x${string}`;
export type Hex = `0x${string}`;

export type EncodeGlobalPaymasterAndDataParams = {
  paymaster: Address;
  npubHash: Hex;
  member: Address;
  /** Member path must use the zero address. */
  policy?: Address;
  verificationGasLimit?: number | bigint;
  postOpGasLimit?: number | bigint;
  version?: number;
};

function strip0x(hex: string): string {
  return hex.startsWith('0x') || hex.startsWith('0X') ? hex.slice(2) : hex;
}

function assertHexLength(label: string, hex: string, byteLength: number): void {
  const raw = strip0x(hex);
  if (!/^[0-9a-fA-F]*$/.test(raw)) {
    throw new Error(`${label} is not hex`);
  }
  if (raw.length !== byteLength * 2) {
    throw new Error(`${label} must be ${byteLength} bytes (got ${raw.length / 2})`);
  }
}

function padUint256(value: number | bigint): string {
  const n = typeof value === 'bigint' ? value : BigInt(value);
  if (n < 0n || n > 2n ** 256n - 1n) {
    throw new Error('value out of uint256 range');
  }
  return n.toString(16).padStart(64, '0');
}

function padAddress(address: Address): string {
  assertHexLength('address', address, 20);
  return strip0x(address).toLowerCase().padStart(64, '0');
}

function encodeUint128(value: number | bigint): string {
  const n = typeof value === 'bigint' ? value : BigInt(value);
  if (n < 0n || n > 2n ** 128n - 1n) {
    throw new Error('value out of uint128 range');
  }
  return n.toString(16).padStart(32, '0');
}

/** abi.encode(uint8, bytes32, address, address) — 128 bytes. */
export function encodeGlobalPaymasterPayload(params: {
  version?: number;
  npubHash: Hex;
  member: Address;
  policy?: Address;
}): Hex {
  const version = params.version ?? GLOBAL_PAYMASTER_DATA_VERSION;
  if (version < 0 || version > 255) {
    throw new Error('version out of uint8 range');
  }
  assertHexLength('npubHash', params.npubHash, 32);
  const policy = params.policy ?? ZERO_ADDRESS;

  const payload =
    padUint256(version) +
    strip0x(params.npubHash).toLowerCase() +
    padAddress(params.member) +
    padAddress(policy);

  return `0x${payload}`;
}

/** Full ERC-4337 `paymasterAndData` (52-byte header + 128-byte payload). */
export function encodeGlobalPaymasterAndData(params: EncodeGlobalPaymasterAndDataParams): Hex {
  assertHexLength('paymaster', params.paymaster, 20);

  const verificationGasLimit =
    params.verificationGasLimit ?? DEFAULT_GLOBAL_PAYMASTER_VERIFICATION_GAS_LIMIT;
  const postOpGasLimit = params.postOpGasLimit ?? DEFAULT_GLOBAL_POST_OP_GAS_LIMIT;

  const header =
    strip0x(params.paymaster).toLowerCase() +
    encodeUint128(verificationGasLimit) +
    encodeUint128(postOpGasLimit);

  const payload = strip0x(
    encodeGlobalPaymasterPayload({
      version: params.version,
      npubHash: params.npubHash,
      member: params.member,
      policy: params.policy,
    }),
  );

  const out = `0x${header}${payload}` as Hex;
  if ((out.length - 2) / 2 !== 180) {
    throw new Error(`unexpected paymasterAndData length ${(out.length - 2) / 2}`);
  }
  return out;
}

/** Pool wei required for a given EntryPoint `maxCost` (115% headroom). */
export function requiredGlobalPoolBalance(maxCostWei: bigint): bigint {
  return (maxCostWei * BigInt(GLOBAL_BALANCE_HEADROOM_BPS)) / 10_000n;
}

File: src-tauri/src/evm/global_paymaster.rs
//! PactoGlobalPaymaster `paymasterAndData` encoding (EntryPoint v0.7).
//! Layout: 52-byte header + abi.encode(uint8 version, bytes32 npubHash, address member, address policy).

use alloy::primitives::{Address, Bytes, B256, U256};
use alloy::sol_types::SolValue;

pub const GLOBAL_PAYMASTER_DATA_VERSION: u8 = 1;
pub const GLOBAL_PAYMASTER_DATA_OFFSET: usize = 52;
pub const GLOBAL_BALANCE_HEADROOM_BPS: u64 = 11_500;
pub const DEFAULT_GLOBAL_PAYMASTER_VERIFICATION_GAS_LIMIT: u128 = 500_000;
pub const DEFAULT_GLOBAL_POST_OP_GAS_LIMIT: u128 = 50_000;

/// Pool wei required for EntryPoint `maxCost` (115% headroom).
pub fn required_global_pool_balance(max_cost_wei: U256) -> U256 {
    max_cost_wei * U256::from(GLOBAL_BALANCE_HEADROOM_BPS) / U256::from(10_000u64)
}

/// Full ERC-4337 `paymasterAndData` (52-byte header + 128-byte ABI payload).
/// Member path: pass `Address::ZERO` for `policy`.
pub fn encode_global_paymaster_and_data(
    paymaster: Address,
    npub_hash: B256,
    member: Address,
    policy: Address,
    verification_gas_limit: u128,
    post_op_gas_limit: u128,
) -> Bytes {
    let mut out = Vec::with_capacity(180);
    out.extend_from_slice(paymaster.as_slice());
    out.extend_from_slice(&verification_gas_limit.to_be_bytes());
    out.extend_from_slice(&post_op_gas_limit.to_be_bytes());
    let payload = (
        U256::from(GLOBAL_PAYMASTER_DATA_VERSION),
        npub_hash,
        member,
        policy,
    )
        .abi_encode();
    out.extend_from_slice(&payload);
    Bytes::from(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloy::primitives::{address, b256};

    #[test]
    fn required_global_pool_balance_matches_golden_headroom() {
        let max = U256::from(1_000_000_000_000_000_000u64);
        assert_eq!(
            required_global_pool_balance(max),
            U256::from(1_150_000_000_000_000_000u64)
        );
    }

    #[test]
    fn encode_global_paymaster_and_data_matches_golden_vector() {
        let paymaster = address!("0x1C2eb4Ac1cD57aF67ad8B20838A28FB23d39d5b8");
        let npub_hash = b256!("0x540d126644e922328318f1870ba0c9de3b2d5c0c271e27af7efea3e44025fdc1");
        let member = address!("0xe05fcC23807536bEe418f142D19fa0d21BB0cfF7");
        let encoded = encode_global_paymaster_and_data(
            paymaster,
            npub_hash,
            member,
            Address::ZERO,
            DEFAULT_GLOBAL_PAYMASTER_VERIFICATION_GAS_LIMIT,
            DEFAULT_GLOBAL_POST_OP_GAS_LIMIT,
        );
        let expected = "0x1c2eb4ac1cd57af67ad8b20838a28fb23d39d5b80000000000000000000000000007a1200000000000000000000000000000c3500000000000000000000000000000000000000000000000000000000000000001540d126644e922328318f1870ba0c9de3b2d5c0c271e27af7efea3e44025fdc1000000000000000000000000e05fcc23807536bee418f142d19fa0d21bb0cff70000000000000000000000000000000000000000000000000000000000000000";
        assert_eq!(format!("{encoded:#x}"), expected);
        assert_eq!(encoded.len(), GLOBAL_PAYMASTER_DATA_OFFSET + 128);
    }
}
